auth/auth/flow.py

```diff
@@ -0,0 +1,123 @@
+import abc
+from typing import Optional
+import msal
+import google.auth.transport.requests
+import google.oauth2.id_token
+import google_auth_oauthlib.flow
+import aiohttp.web
+import json
+import urllib.parse
+
+from gear.cloud_config import get_global_config
+
+
+class FlowData:
+    def __init__(self, authorization_url: str, state: str):
+        self.authorization_url = authorization_url
+        self.state = state
+
+
+class FlowResult:
+    def __init__(self, login_id: str, email: str, token: dict):
+        self.login_id = login_id
+        self.email = email
+        self.token = token
+
+
+class Flow(abc.ABC):
+    @abc.abstractmethod
+    def initialize_flow(self) -> FlowData:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def finish_flow(self, request: aiohttp.web.Request, flow_dict: Optional[dict]) -> FlowResult:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def as_dict(self):
+        raise NotImplementedError
+
+
+class GoogleFlow(Flow):
+    @staticmethod
+    def get_flow(credentials_file: str, redirect_uri: str, state: Optional[str] = None) -> 'GoogleFlow':
+        scopes = [
+            'https://www.googleapis.com/auth/userinfo.profile',
+            'https://www.googleapis.com/auth/userinfo.email',
+            'openid',
+        ]
+
+        flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(credentials_file, scopes=scopes, state=state)
+        flow.redirect_uri = redirect_uri
+
+        return GoogleFlow(flow)
+
+    def __init__(self, flow: google_auth_oauthlib.flow.Flow):
+        self._flow = flow
+
+    def initialize_flow(self) -> FlowData:
+        authorization_url, state = self._flow.authorization_url(access_type='offline', include_granted_scopes='true')
+        return FlowData(authorization_url, state)
+
+    def finish_flow(
+        self, request: aiohttp.web.Request, flow_dict: Optional[dict]
+    ) -> FlowResult:  # pylint: disable=unused-argument
+        self._flow.fetch_token(code=request.query['code'])
+        token = google.oauth2.id_token.verify_oauth2_token(
+            self._flow.credentials.id_token, google.auth.transport.requests.Request()
+        )
+        email = token['email']
+        return FlowResult(email, email, token)
+
+    def as_dict(self) -> Optional[dict]:
+        return None
+
+
+class AzureFlow(Flow):
+    @staticmethod
+    def get_flow(credentials_file: str, redirect_uri: str, state: Optional[str] = None) -> 'AzureFlow':
+        with open(credentials_file) as f:
+            data = json.loads(f.read())
+
+        tenant_id = data['tenant']
+        authority = f'https://login.microsoftonline.com/{tenant_id}'
+        client = msal.ConfidentialClientApplication(data['appId'], data['password'], authority)
+
+        flow = client.initiate_auth_code_flow(scopes=[], redirect_uri=redirect_uri, state=state)
```

Comment:
This should happen in `initialize_flow`, not in this constructor. We don't want to initiate an auth flow the second time we call `get_flow` in the oauth2callback.